File: java/integTest/src/test/java/glide/DnsTest.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide;

import static glide.TestUtilities.*;
import static org.junit.jupiter.api.Assertions.*;

import glide.api.BaseClient;
import glide.api.GlideClient;
import glide.api.GlideClusterClient;
import glide.api.models.configuration.AdvancedGlideClientConfiguration;
import glide.api.models.configuration.AdvancedGlideClusterClientConfiguration;
import glide.api.models.configuration.GlideClientConfiguration;
import glide.api.models.configuration.GlideClientConfiguration.GlideClientConfigurationBuilder;
import glide.api.models.configuration.GlideClusterClientConfiguration;
import glide.api.models.configuration.GlideClusterClientConfiguration.GlideClusterClientConfigurationBuilder;
import glide.api.models.configuration.NodeAddress;
import glide.api.models.configuration.TlsAdvancedConfiguration;
import lombok.SneakyThrows;
import org.junit.jupiter.api.Timeout;
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

/**
 * Tests for client behavior when connecting with hostnames.
 *
 * <p>To run these tests, you need to add the following mappings to your hosts file then set the
 * environment variable {@code VALKEY_GLIDE_DNS_TESTS_ENABLED}:
 *
 * <ul>
 *   <li>{@code 127.0.0.1 valkey.glide.test.tls.com}
 *   <li>{@code 127.0.0.1 valkey.glide.test.no_tls.com}
 *   <li>{@code ::1 valkey.glide.test.tls.com}
 *   <li>{@code ::1 valkey.glide.test.no_tls.com}
 * </ul>
 */
@Timeout(10)
@EnabledIfEnvironmentVariable(named = "VALKEY_GLIDE_DNS_TESTS_ENABLED", matches = ".*")
public class DnsTest {

    // Hostname constants for testing. Both hostnames should map to localhost,
    // but only HOSTNAME_TLS should be included in the TLS certificate used by the
    // server. See 'cluster_manager.py' for details.
    private static final String HOSTNAME_TLS = "valkey.glide.test.tls.com";
    private static final String HOSTNAME_NO_TLS = "valkey.glide.test.no_tls.com";

    @ParameterizedTest
    @ValueSource(booleans = {true, false})
    @SneakyThrows
    void testConnectWithValidHostnameNoTls(boolean clusterMode) {
        try (BaseClient client = buildClient(clusterMode, false, HOSTNAME_NO_TLS)) {
            assertConnected(client);
        }
    }

    @ParameterizedTest
    @ValueSource(booleans = {true, false})
    void testConnectWithInvalidHostnameNoTls(boolean clusterMode) {
        assertThrows(Exception.class, () -> buildClient(clusterMode, false, "nonexistent.invalid"));
    }

    @ParameterizedTest
    @ValueSource(booleans = {true, false})
    @SneakyThrows
    void testTlsConnectWithHostnameInCert(boolean clusterMode) {
        try (BaseClient client = buildClient(clusterMode, true, HOSTNAME_TLS)) {
            assertConnected(client);
        }
    }

    @ParameterizedTest
    @ValueSource(booleans = {true, false})
    void testTlsConnectWithHostnameNotInCert(boolean clusterMode) {
        assertThrows(Exception.class, () -> buildClient(clusterMode, true, HOSTNAME_NO_TLS));
    }

    // -------------------
    // Helper methods
    // -------------------

    @SneakyThrows
    private static BaseClient buildClient(boolean clusterMode, boolean useTls, String hostname) {

        // Get port from test configuration.
        String[] hosts =
                useTls
                        ? (clusterMode
                                ? TestConfiguration.CLUSTER_TLS_HOSTS
                                : TestConfiguration.STANDALONE_TLS_HOSTS)
                        : (clusterMode ? TestConfiguration.CLUSTER_HOSTS : TestConfiguration.STANDALONE_HOSTS);
        String hostEntry = hosts[0].trim();
        int port = Integer.parseInt(hostEntry.substring(hostEntry.lastIndexOf(':') + 1));

        // Build common arguments.
        NodeAddress address = NodeAddress.builder().host(hostname).port(port).build();
        byte[] certificateBytes = getCaCertificate();

        // Build client configuration and client.
        if (clusterMode) {
            GlideClusterClientConfigurationBuilder<?, ?> builder =
                    GlideClusterClientConfiguration.builder().address(address).useTLS(useTls);
            if (useTls) {
                TlsAdvancedConfiguration tlsConfig =
                        TlsAdvancedConfiguration.builder().rootCertificates(certificateBytes).build();
                builder.advancedConfiguration(
                        AdvancedGlideClusterClientConfiguration.builder()
                                .tlsAdvancedConfiguration(tlsConfig)
                                .build());
            }
            return GlideClusterClient.createClient(builder.build()).get();
        } else {
            GlideClientConfigurationBuilder<?, ?> builder =
                    GlideClientConfiguration.builder().address(address).useTLS(useTls);
            if (useTls) {
                TlsAdvancedConfiguration tlsConfig =
                        TlsAdvancedConfiguration.builder().rootCertificates(certificateBytes).build();
                builder.advancedConfiguration(
                        AdvancedGlideClientConfiguration.builder().tlsAdvancedConfiguration(tlsConfig).build());
            }
            return GlideClient.createClient(builder.build()).get();
        }
    }
}

File: node/tests/Dns.test.ts
/**
 * Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
 */

import { afterAll, beforeAll, describe, expect, it } from "@jest/globals";
import { TestTLSConfig, ValkeyCluster } from "../../utils/TestUtils.js";
import { GlideClient, GlideClusterClient, ProtocolVersion } from "../build-ts";
import { HOSTNAME_NO_TLS, HOSTNAME_TLS } from "./Constants";
import {
    assertConnected,
    getCaCertificateData,
    getClientConfigurationOption,
    getServerVersion,
} from "./TestUtilities";

const TIMEOUT = 50000;
const CLUSTER_CREATION_TIMEOUT = 120000;
const HOSTNAME_INVALID = "nonexistent.invalid";

/**
 * Returns true if DNS tests are enabled via the VALKEY_GLIDE_DNS_TESTS_ENABLED environment variable.
 */
const isDnsTestsEnabled = () =>
    process.env.VALKEY_GLIDE_DNS_TESTS_ENABLED !== undefined;

/**
 * Creates base configuration for connecting to a server with the specified hostname.
 *
 * @param server - The ValkeyCluster server instance
 * @param hostname - The hostname to connect to
 * @returns Base client configuration
 */
function createBaseConfig(server: ValkeyCluster, hostname: string) {
    const port = server.ports()[0];
    const address = [hostname, port] as [string, number];
    return {
        ...getClientConfigurationOption([address], ProtocolVersion.RESP3),
    };
}

/**
 * Creates a standalone client connected to the given server with the specified hostname.
 *
 * @param server - The ValkeyCluster server instance
 * @param hostname - The hostname to connect to
 * @returns A connected GlideClient instance
 */
async function createClient(
    server: ValkeyCluster,
    hostname: string,
): Promise<GlideClient> {
    const baseConfig = createBaseConfig(server, hostname);

    if (server.isTls()) {
        return await GlideClient.createClient({
            ...baseConfig,
            useTLS: true,
            advancedConfiguration: {
                tlsAdvancedConfiguration: {
                    rootCertificates: getCaCertificateData(),
                },
            },
        });
    }

    return await GlideClient.createClient(baseConfig);
}

/**
 * Creates a cluster client connected to the given server with the specified hostname.
 *
 * @param server - The ValkeyCluster server instance
 * @param hostname - The hostname to connect to
 * @returns A connected GlideClusterClient instance
 */
async function createClusterClient(
    server: ValkeyCluster,
    hostname: string,
): Promise<GlideClusterClient> {
    const baseConfig = createBaseConfig(server, hostname);

    if (server.isTls()) {
        return await GlideClusterClient.createClient({
            ...baseConfig,
            useTLS: true,
            advancedConfiguration: {
                tlsAdvancedConfiguration: {
                    rootCertificates: getCaCertificateData(),
                },
            },
        });
    }

    return await GlideClusterClient.createClient(baseConfig);
}

(isDnsTestsEnabled() ? describe : describe.skip)("DNS Tests - Non-TLS", () => {
    let standaloneServer: ValkeyCluster;
    let clusterServer: ValkeyCluster;

    beforeAll(async () => {
        standaloneServer = await ValkeyCluster.createCluster(
            false,
            1,
            1,
            getServerVersion,
        );
        clusterServer = await ValkeyCluster.createCluster(
            true,
            3,
            2,
            getServerVersion,
        );
    }, CLUSTER_CREATION_TIMEOUT);

    afterAll(async () => {
        await standaloneServer.close();
        await clusterServer.close();
    });

    it(
        "should connect with valid hostname - standalone",
        async () => {
            const client = await createClient(
                standaloneServer,
                HOSTNAME_NO_TLS,
            );

            await assertConnected(client);
            client.close();
        },
        TIMEOUT,
    );

    it(
        "should fail with invalid hostname - standalone",
        async () => {
            await expect(
                createClient(standaloneServer, HOSTNAME_INVALID),
            ).rejects.toThrow();
        },
        TIMEOUT,
    );

    it(
        "should connect with valid hostname - cluster",
        async () => {
            const client = await createClusterClient(
                clusterServer,
                HOSTNAME_NO_TLS,
            );

            await assertConnected(client);
            client.close();
        },
        TIMEOUT,
    );

    it(
        "should fail with invalid hostname - cluster",
        async () => {
            await expect(
                createClusterClient(clusterServer, HOSTNAME_INVALID),
            ).rejects.toThrow();
        },
        TIMEOUT,
    );
});

(isDnsTestsEnabled() ? describe : describe.skip)("DNS Tests - TLS", () => {
    let standaloneServer: ValkeyCluster;
    let clusterServer: ValkeyCluster;

    beforeAll(async () => {
        const tlsConfig: TestTLSConfig = {
            useTLS: true,
            advancedConfiguration: {
                tlsAdvancedConfiguration: {
                    insecure: true,
                },
            },
        };

        standaloneServer = await ValkeyCluster.createCluster(
            false,
            1,
            1,
            getServerVersion,
            true,
            tlsConfig,
        );
        clusterServer = await ValkeyCluster.createCluster(
            true,
            3,
            2,
            getServerVersion,
            true,
            tlsConfig,
        );
    }, CLUSTER_CREATION_TIMEOUT);

    afterAll(async () => {
        await standaloneServer.close();
        await clusterServer.close();
    });

    it(
        "should connect with hostname in certificate SAN - standalone",
        async () => {
            const client = await createClient(standaloneServer, HOSTNAME_TLS);

            await assertConnected(client);
            client.close();
        },
        TIMEOUT,
    );

    it(
        "should fail with hostname NOT in certificate SAN - standalone",
        async () => {
            await expect(
                createClient(standaloneServer, HOSTNAME_NO_TLS),
            ).rejects.toThrow();
        },
        TIMEOUT,
    );

    it(
        "should connect with hostname in certificate SAN - cluster",
        async () => {
            const client = await createClusterClient(
                clusterServer,
                HOSTNAME_TLS,
            );

            await assertConnected(client);
            client.close();
        },
        TIMEOUT,
    );

    it(
        "should fail with hostname NOT in certificate SAN - cluster",
        async () => {
            await expect(
                createClusterClient(clusterServer, HOSTNAME_NO_TLS),
            ).rejects.toThrow();
        },
        TIMEOUT,
    );
});
